src/concurrency/timestamp_ordering_transaction_manager.cpp

```diff
@@ -397,83 +163,188 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
 // It should not be called if the tuple is in the write set as commit and abort
 // will release the write lock anyway.
 void TimestampOrderingTransactionManager::YieldOwnership(
-    UNUSED_ATTRIBUTE Transaction *const current_txn, const oid_t &tile_group_id,
+    UNUSED_ATTRIBUTE Transaction *const current_txn, 
+    const storage::TileGroupHeader *const tile_group_header,
     const oid_t &tuple_id) {
-  auto &manager = catalog::Manager::GetInstance();
-  auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
   PL_ASSERT(IsOwner(current_txn, tile_group_header, tuple_id));
   tile_group_header->SetTransactionId(tuple_id, INITIAL_TXN_ID);
 }
 
 bool TimestampOrderingTransactionManager::PerformRead(
     Transaction *const current_txn, const ItemPointer &location,
     bool acquire_ownership) {
-  if (current_txn->IsDeclaredReadOnly() == true) {
-    // Ignore read validation for all readonly transactions
+  //////////////////////////////////////////////////////////
+  //// handle READ_ONLY
+  //////////////////////////////////////////////////////////
+  if (current_txn->GetIsolationLevel() == IsolationLevelType::READ_ONLY) {
+    // do not update read set for read-only transactions.
     return true;
-  }
+  } // end READ ONLY
+
+  //////////////////////////////////////////////////////////
+  //// handle SNAPSHOT
+  //////////////////////////////////////////////////////////
+  else if (current_txn->GetIsolationLevel() == IsolationLevelType::SNAPSHOT) {
+    // Check if it's select for update before we check the ownership 
+    // and modify the last reader tid
+    if (acquire_ownership == true) {
+    
+      oid_t tile_group_id = location.block;
+      oid_t tuple_id = location.offset;
+
+      LOG_TRACE("PerformRead (%u, %u)\n", location.block, location.offset);
+      auto &manager = catalog::Manager::GetInstance();
+      auto tile_group = manager.GetTileGroup(tile_group_id);
+      auto tile_group_header = tile_group->GetHeader();
+
+      if (IsOwner(current_txn, tile_group_header, tuple_id) == false) {
+        // Acquire ownership if we haven't
+        if (IsOwnable(current_txn, tile_group_header, tuple_id) == false) {
+          // Cannot own
+          return false;
+        }
+        if (AcquireOwnership(current_txn, tile_group_header, tuple_id) == false) {
+          // Cannot acquire ownership
+          return false;
+        }
+        // Promote to RWType::READ_OWN
+        current_txn->RecordReadOwn(location);
+      }
+      // if the ownership has already been acquired, 
+      // then no need to update read/write set.
+      return true;
+    
+    } else {
+      // if it's not select for update, then directly return true.
+      // no need to update read/write set.
+      return true;
```

Comment:
Please read the original snapshot isolation paper https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-95-51.pdf first, and then we discuss offline.